pkg/aws/awscontrolplane/validate_awscontrolplane.go

```diff
@@ -43,18 +43,98 @@ func NewValidator(config config.Config) (*Validator, error) {
 
 func (v *Validator) Validate(request *v1beta1.AdmissionRequest) (bool, error) {
 	var awsControlPlane infrastructurev1alpha2.AWSControlPlane
-	var allowed bool
 
 	if _, _, err := validator.Deserializer.Decode(request.Object.Raw, nil, &awsControlPlane); err != nil {
 		return false, microerror.Maskf(aws.ParsingFailedError, "unable to parse awscontrol plane: %v", err)
 	}
-	allowed, err := v.ControlPlaneLabelMatch(awsControlPlane)
+	controlPlaneLabelMatches, err := v.ControlPlaneLabelMatch(awsControlPlane)
 	if err != nil {
 		return false, microerror.Mask(err)
 
 	}
+	azCountAllowed, err := v.AZCount(awsControlPlane)
+	if err != nil {
+		return false, microerror.Mask(err)
+
+	}
+	azReplicaMatches, err := v.AZReplicaMatch(awsControlPlane)
+	if err != nil {
+		return false, microerror.Mask(err)
+
+	}
+
+	return controlPlaneLabelMatches && azCountAllowed && azReplicaMatches, nil
+}
+
+func (v *Validator) AZReplicaMatch(awsControlPlane infrastructurev1alpha2.AWSControlPlane) (bool, error) {
+	var g8sControlPlane infrastructurev1alpha2.G8sControlPlane
+	var err error
+	var fetch func() error
+
+	// Fetch the G8sControlPlane.
+	{
+		v.Log("level", "debug", "message", fmt.Sprintf("Fetching G8sControlPlane %s", awsControlPlane.Name))
+		fetch = func() error {
+			ctx := context.Background()
+
+			err = v.k8sClient.CtrlClient().Get(
+				ctx,
+				types.NamespacedName{Name: awsControlPlane.GetName(), Namespace: awsControlPlane.GetNamespace()},
+				&g8sControlPlane,
+			)
+			if err != nil {
+				return microerror.Maskf(aws.NotFoundError, "failed to fetch g8sControlplane: %v", err)
+			}
+			return nil
+		}
+	}
+
+	{
+		b := backoff.NewMaxRetries(3, 1*time.Second)
+		err = backoff.Retry(fetch, b)
+		// Note that while we do log the error, we don't fail if the G8sControlPlane doesn't exist yet. That is okay because the order of CR creation can vary.
+		if aws.IsNotFound(err) {
+			v.Log("level", "debug", "message", fmt.Sprintf("No G8sControlPlane %s could be found: %v", awsControlPlane.GetName(), err))
+			return true, nil
+		} else if err != nil {
+			return false, microerror.Mask(err)
+		}
+	}
+
+	if g8sControlPlane.Spec.Replicas != len(awsControlPlane.Spec.AvailabilityZones) {
+		v.logger.Log("level", "debug", "message", fmt.Sprintf("G8sControlPlane %s with %v replicas does not match AWSControlPlane %s with %v availability zones %s",
```

Comment:
Why are we doing logging and error handling at the same time? To me it always appeared to be redundant.

Reply:
IIRC there were some arguments for this when it was introduced elsewhere. Personally I have no preference but I kept it in the same format as the other handlers (logging and then throwing an error)